crates/containerd-shim-wasmer/src/executor.rs

```diff
@@ -0,0 +1,123 @@
+use containerd_shim_wasm::sandbox::oci::{self, Spec};
+use containerd_shim_wasm::sandbox::Stdio;
+use libcontainer::workload::{Executor, ExecutorError, ExecutorValidationError};
+use wasmer::{Cranelift, Module, Store};
+use wasmer_wasix::WasiEnv;
+
+use crate::oci_wasmer;
+
+const EXECUTOR_NAME: &str = "wasmer";
+
+#[derive(Clone)]
+pub struct WasmerExecutor {
+    stdio: Stdio,
+    engine: Cranelift,
+}
+
+impl WasmerExecutor {
+    pub fn new(stdio: Stdio, engine: Cranelift) -> Self {
+        Self { stdio, engine }
+    }
+}
+
+impl Executor for WasmerExecutor {
+    fn exec(&self, spec: &Spec) -> Result<(), ExecutorError> {
+        match can_handle(spec) {
+            Ok(()) => {
+                let args = oci::get_args(spec);
+                if args.is_empty() {
+                    return Err(ExecutorError::InvalidArg);
+                }
+
+                match self.start(spec, args) {
+                    Ok(_) => std::process::exit(0),
+                    Err(_e) => std::process::exit(137),
+                }
+            }
+            Err(ExecutorValidationError::CantHandle(_)) => {
+                LinuxContainerExecutor::new(self.stdio.clone()).exec(spec)?;
+
+                Ok(())
+            }
+            Err(_) => Err(ExecutorError::InvalidArg),
+        }
+    }
+
+    fn validate(&self, spec: &Spec) -> std::result::Result<(), ExecutorValidationError> {
+        match can_handle(spec) {
+            Ok(()) => Ok(()),
+            Err(ExecutorValidationError::CantHandle(_)) => {
+                LinuxContainerExecutor::new(self.stdio.clone()).validate(spec)?;
+
+                Ok(())
+            }
+            Err(err) => Err(err),
+        }
+    }
+}
+
+impl WasmerExecutor {
+    fn start(&self, spec: &Spec, args: &[String]) -> anyhow::Result<()> {
+        log::info!("get envs from spec");
+        let envs = oci_wasmer::env_to_wasi(spec);
+
+        log::info!("redirect stdio");
+        self.stdio.take().redirect()?;
+
+        log::info!("get module_name and method");
+        let (module_name, _method) = oci::get_module(spec);
+        let module_name = match module_name {
+            Some(m) => m,
+            None => {
+                return Err(anyhow::format_err!(
+                    "no module provided, cannot load module from file within container"
+                ))
+            }
+        };
+
+        log::info!("Create a Store");
+        let mut store = Store::new(self.engine.clone());
+
+        log::info!("loading module from file {} ", module_name);
+        let module = Module::from_file(&store, module_name)?;
+
+        log::info!("Run the module args: {:?}, envs: {:?}", args, envs);
+        WasiEnv::builder(EXECUTOR_NAME)
+            .args(args[1..].to_vec())
+            .envs(envs)
+            .preopen_dir("/")?
+            .run_with_store(module, &mut store)?;
+
+        Ok(())
+    }
+}
+
+fn can_handle(spec: &Spec) -> Result<(), ExecutorValidationError> {
```

Comment:
Wow, that seems like it could be done.

There are two places that use this now: https://github.com/containerd/runwasi/blob/7e978edeaf34b54efb895738357c805cf888b76d/crates/containerd-shim-wasmtime/src/executor.rs#L29-L31

https://github.com/containerd/runwasi/blob/7e978edeaf34b54efb895738357c805cf888b76d/crates/containerd-shim-wasmtime/src/executor.rs#L56-L58

Create open a PR to fix it?

Reply:
This will be fixed by #293.